rfcs/0001-masonry-backend.md

```diff
@@ -0,0 +1,166 @@
+
+Xilem is roughly separated into two layers:
+
+- **The reactive layer** handles the app logic and centralized state. Your application's code will call functions from this layer.
+- **The platform layer** handles the document state, depending on the platform.
+
+The reactive layer defines the `View` trait as well as other items for sequences, type-erasure and memoization. A `View` represent a transient node or tree of graphical items, which is return by a declarative function we usually call a "component". An example of component could be:
+
+```rust
+fn app_logic(data: &mut u32) -> impl View<u32, (), Element = impl Widget> {
+    Column::new((
+        Button::new(format!("count: {}", data), |data| *data += 1),
+        Button::new("reset", |data| *data = 0),
+    ))
+}
+```
+
+In that example, `app_logic()` returns a `Column` view which contains two `Button` views. These store the minimum information to represent the concept of two buttons stored in a column.
+
+The `View` trait has an `Element` associated type, which depends on the platform layer, and represents a persisten UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no longer has a child view that was in the previous version, the matching element is removed, and so on.
+
+(In web frameworks that use a virtual DOM, this process is sometimes known as "reconciliation".)
+
+Xilem has multiple platform layers. `xilem_web` is based on the web DOM, whereas `xilem_native` is based on the Masonry crate, which provides a native widget tree and some Developer Experience features.
+
+Note that, while the platform layer is sometimes referred to as a "backend", it's somewhat tightly coupled with the reactive layer.
+
+`xilem_web`, `xilem_native` and any other platforms will each have their own version of the `View` trait, the `ViewSequence` trait, the `AnyView` trait, the `BoxedView` type, and so on.
+
+### `xilem_native`
+
+The element type in `xilem_native` must implement `masonry::Widget`.
+
+Widget in Masonry is a trait that represent units of graphical composition, things like text and buttons and boxes and containers. To implement Widget, a type must:
+
+- Be `'static`.
+- Provide info about its children.
+- Handle user events.
+- Compute its layout.
+- Draw itself in a paint target.
+
+See the [Widget trait definition](https://github.com/PoignardAzur/masonry-rs/blob/3cfc2cddec05d48a31a9812bdda27792d336f03f/src/widget/widget.rs#L65-L199) for more details.
+
+
+### Directory structure
+
+The `https://github.com/linebender/xilem` repository has the following crates in its `crates/` directory:
+
+- `xilem_core/`
+- `xilem_web/`
+- `xilem_native/`
+- `masonry/`
+- `xilem/`, which would only be a placeholder crate pointing to `xilem_web/` or `xilem_native/`.
```

Comment:
Pointing to here meaning:
a) A link in the readme, and no content
b) Re-exporting only `xilem_native`
c) Re-exporting `xilem_native` or `xilem_web`, depending on platform
?

Reply:
There are tradeoffs here. My preference in the near term would be for `xilem` itself to be the crate which users depend on for the native backend, with no need to know about `xilem_native`.

However, that does constrain the longer-term web/native backend unification/abstraction, which I hope might happen, to not use `xilem` as the crate name, which would be unfortunate.